cmd/insights/commands/upload.go

```diff
@@ -40,23 +42,45 @@ func installUploadCmd(app *App) {
 	uploadCmd.Flags().UintVar(&app.config.Upload.MinAge, "min-age", defaultMinAge, "the minimum age (in seconds) of a report before the uploader will attempt to upload it")
 	uploadCmd.Flags().BoolVarP(&app.config.Upload.Force, "force", "f", false, "force an upload, ignoring min age and clashes between the collected file and a file in the uploaded folder, replacing the clashing uploaded report if it exists")
 	uploadCmd.Flags().BoolVarP(&app.config.Upload.DryRun, "dry-run", "d", false, "go through the motions of doing an upload, but do not communicate with the server or send the payload")
+	uploadCmd.Flags().BoolVarP(&app.config.Upload.Retry, "retry", "r", false, "enable a limited number of retries for failed uploads")
 
 	app.cmd.AddCommand(uploadCmd)
 }
 
 func (a App) uploadRun() error {
 	cm := consent.New(a.config.consentDir)
 
+	var wg sync.WaitGroup
+	uploaders := make(map[string]uploader.Uploader)
 	for _, source := range a.config.Upload.Sources {
-		u, err := a.newUploader(cm, a.config.insightsDir, source, a.config.Upload.MinAge, a.config.Upload.DryRun)
+		u, err := a.newUploader(cm, a.config.insightsDir, source, a.config.Upload.MinAge, a.config.Upload.DryRun, a.config.Upload.Retry)
 		if err != nil {
 			return fmt.Errorf("failed to create uploader for source %s: %v", source, err)
 		}
-
-		if err := u.Upload(a.config.Upload.Force); err != nil {
-			return fmt.Errorf("failed to upload reports for source %s: %v", source, err)
-		}
+		uploaders[source] = u
 	}
 
-	return nil
+	var uploadError error
+	mu := &sync.Mutex{}
```

Comment:
Would it make sense to move `wg`'s declaration to near the mutex since it isn't used before?